In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import FinanceDataReader as fdr
from datetime import datetime
import seaborn as sns
import statsmodels.api as sm

%matplotlib inline

* US SP MTK CAP DATA RESOURCE : Joe Abbott (yardeni researcher)
https://www.yardeni.com/
* 단위 : Billion

In [4]:
path = '/Users/taehyeonkang/pythonfiles/data/삼성전자회귀분석(2010-2017)'

In [5]:
cap_df = pd.read_excel(path+'/JA20180819_DailySPMarketCaps-BrianKang.xlsx', sheet_name='Sheet1')

In [8]:
cap_df.head(2)

Market Cap (billion US$) S&P 1500  S&P 500 S&P 400 S&P 600
0              Dec 30 1994  3987.98  3346.12  462.34  179.52
1              Jan 02 1995      ---      ---     ---     ---

In [6]:
cap_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6166 entries, 0 to 6165
Data columns (total 5 columns):
Market Cap (billion US$)    6166 non-null object
S&P 1500                    6166 non-null object
S&P 500                     6166 non-null object
S&P 400                     6166 non-null object
S&P 600                     6166 non-null object
dtypes: object(5)
memory usage: 241.0+ KB


In [9]:
cap_df.head()

Market Cap (billion US$) S&P 1500  S&P 500 S&P 400 S&P 600
0              Dec 30 1994  3987.98  3346.12  462.34  179.52
1              Jan 02 1995      ---      ---     ---     ---
2              Jan 03 1995  3981.99  3345.01  459.92  177.06
3              Jan 04 1995  3995.66  3356.64  461.75  177.27
4              Jan 05 1995  3992.05  3353.93  460.84  177.28

In [16]:
df_col = cap_df.columns.values
df_col[0] = 'date'
cap_df.columns = df_col
cap_df = cap_df.set_index('date')

In [17]:
cap_df.index = pd.to_datetime(cap_df.index)

In [18]:
cap_df.head()

S&P 1500  S&P 500 S&P 400 S&P 600
date                                        
1994-12-30  3987.98  3346.12  462.34  179.52
1995-01-02      ---      ---     ---     ---
1995-01-03  3981.99  3345.01  459.92  177.06
1995-01-04  3995.66  3356.64  461.75  177.27
1995-01-05  3992.05  3353.93  460.84  177.28

In [19]:
cap_df.columns.tolist()

['S&P 1500', 'S&P 500', 'S&P 400', 'S&P 600']

In [20]:
cap_df = cap_df.apply(pd.to_numeric, errors = 'coerce')

In [21]:
len(cap_df['2010':'2017'])

2086

In [22]:
len(cap_df['2010':'2017'].dropna())

2012

In [23]:
spINX = cap_df['2010':'2017'].dropna().copy()

In [24]:
spINX.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 2012 entries, 2010-01-04 to 2017-12-29
Data columns (total 4 columns):
S&P 1500    2012 non-null float64
S&P 500     2012 non-null float64
S&P 400     2012 non-null float64
S&P 600     2012 non-null float64
dtypes: float64(4)
memory usage: 78.6 KB


In [25]:
spINX.head()

S&P 1500   S&P 500  S&P 400  S&P 600
date                                            
2010-01-04  11416.01  10088.69   917.16   410.16
2010-01-05  11448.55  10120.19   919.60   408.75
2010-01-06  11458.48  10125.66   924.45   408.36
2010-01-07  11506.43  10166.24   929.12   411.06
2010-01-08  11542.36  10195.52   934.47   412.37

In [26]:
df = pd.read_csv(path+'/sp500회귀분석기초데이터.csv', index_col=0) # 단위 : Million
df.head()

AAPL_CAP      AAPL  SP500_point     SP500_CAP     SP500  \
Date                                                                       
2010-01-04            NaN       NaN      1132.99           NaN       NaN   
2010-01-05  155650.970360  0.001963      1136.52  1.028779e+07  0.003116   
2010-01-06  153160.961366 -0.015997      1137.14  1.029340e+07  0.000546   
2010-01-07  152856.062306 -0.001991      1141.69  1.033459e+07  0.004001   
2010-01-08  153872.392507  0.006649      1144.98  1.036437e+07  0.002882   

            SP500-AAPL(CAP)  NewSP500  
Date                                   
2010-01-04              NaN       NaN  
2010-01-05     1.013214e+07       NaN  
2010-01-06     1.014024e+07  0.000800  
2010-01-07     1.018173e+07  0.004092  
2010-01-08     1.021049e+07  0.002825

In [27]:
df.tail()

AAPL_CAP      AAPL  SP500_point     SP500_CAP     SP500  \
Date                                                                       
2017-12-22  889339.742823  0.000000      2683.34  2.428961e+07 -0.000458   
2017-12-26  866777.212350 -0.025370      2680.50  2.426390e+07 -0.001058   
2017-12-27  866929.661880  0.000176      2682.62  2.428310e+07  0.000791   
2017-12-28  869368.854364  0.002814      2687.54  2.432763e+07  0.001834   
2017-12-29  859967.800000 -0.010814      2673.61  2.420154e+07 -0.005183   

            SP500-AAPL(CAP)  NewSP500  
Date                                   
2017-12-22     2.340027e+07 -0.000476  
2017-12-26     2.339713e+07 -0.000134  
2017-12-27     2.341617e+07  0.000814  
2017-12-28     2.345826e+07  0.001798  
2017-12-29     2.334157e+07 -0.004975

In [28]:
print("2010-01-05, 역산한 값 : %5.3f" % (df['SP500_CAP']['2010-01-05']*1000000/1000000000), "US Billion")
print('2010-01-05, 제공받은 값(yardeni) : %5.3f' % spINX['S&P 500']['2010-01-05'], 'US Billion')
print("")
print("2017-12-29, 제공받은 값 : %5.3f" % (df['SP500_CAP']['2017-12-29']*1000000/1000000000), "US Billion")
print('2017-12-29, 제공받은 값(yardeni) : %5.3f' % spINX['S&P 500']['2017-12-29'], 'US Billion')

2010-01-05, 역산한 값 : 10287.787 US Billion
2010-01-05, 제공받은 값(yardeni) : 10120.190 US Billion

2017-12-29, 제공받은 값 : 24201.537 US Billion
2017-12-29, 제공받은 값(yardeni) : 22821.240 US Billion


In [29]:
24201.537 - 22821.240

1380.2969999999987

* yardeni의 데이터의 단위 : Billion (출처 : 메일로 받음)
* siblisresearch의 데이터 단위 : Million (출처 : http://siblisresearch.com/data/market-caps-sp-100-us/)



* 2017-12-29 서로 다른 리서치 기관에서 제공한 두 값은 약 1380 B 차이남
* 2017-12-29 동일 대상인 S&P 500 시가총액 데이터의 값이 다른 이유는 단위표기의 차이에서 발생하는 오차로 인한 것?

In [30]:
aaplReturn = fdr.DataReader('AAPL', '2009-12', '2017-12-31')
aaplReturn.head()

Close  Open  High   Low       Volume  Change
Date                                                    
2009-12-01   7.03  7.22  7.24  7.03  465760000.0 -0.0154
2009-12-02   7.01  7.11  7.19  6.99  715260000.0 -0.0028
2009-12-03   7.02  7.05  7.11  7.01  448720000.0  0.0014
2009-12-04   6.90  7.13  7.14  6.80  827110000.0 -0.0171
2009-12-07   6.75  6.90  6.92  6.74  714760000.0 -0.0217

In [31]:
aaplReturn.tail()

Close   Open   High    Low       Volume  Change
Date                                                       
2017-12-22  43.75  43.67  43.86  43.62   65400000.0  0.0000
2017-12-26  42.64  42.70  42.87  42.42  132740000.0 -0.0254
2017-12-27  42.65  42.52  42.70  42.43   85990000.0  0.0002
2017-12-28  42.77  42.75  42.96  42.62   65920000.0  0.0028
2017-12-29  42.31  42.63  42.65  42.30  104000000.0 -0.0108

In [270]:
aapl = pd.DataFrame()
aapl['aapl'] = aaplReturn['Close'] / aaplReturn['Close'].shift(1) - 1
aapl['2009-12-31':'2017'].tail()

aapl
Date                
2017-12-22  0.000000
2017-12-26 -0.025370
2017-12-27  0.000176
2017-12-28  0.002814
2017-12-29 -0.010814

In [271]:
aaplcap = pd.DataFrame()
aaplcap['return'] = aapl['aapl']['2009-12-31':]

In [272]:
aaplcap['cap'] = 859967.8

In [273]:
aaplcap['cumreturn'] = aaplcap['return'].add(1).cumprod()

In [274]:
lastvalue = aaplcap['cumreturn'][-1]

In [275]:
aaplcap['cumreturn'] = aaplcap['cumreturn']/lastvalue

In [276]:
aaplcap['cap'] = 859967.8 * aaplcap['cumreturn']

In [277]:
aaplcap.head()

return            cap  cumreturn
Date                                          
2009-12-31 -0.004300  152957.695326   0.177864
2010-01-04  0.015615  155346.071299   0.180642
2010-01-05  0.001963  155650.970360   0.180996
2010-01-06 -0.015997  153160.961366   0.178101
2010-01-07 -0.001991  152856.062306   0.177746

In [233]:
aaplcap.tail()

return            cap  cumreturn
Date                                          
2017-12-22  0.000000  889339.742823   1.034155
2017-12-26 -0.025370  866777.212350   1.007918
2017-12-27  0.000176  866929.661880   1.008095
2017-12-28  0.002814  869368.854364   1.010932
2017-12-29 -0.010814  859967.800000   1.000000

In [259]:
len(aaplcap)

2015

In [261]:
new_df = pd.DataFrame()
new_df['SP500 CAP'] = spINX['S&P 500'] # 단위 : 빌리언
new_df['SP400 CAP'] = spINX['S&P 400'] # 단위 : 빌리언
new_df['SP600 CAP'] = spINX['S&P 600'] # 단위 : 빌리언
new_df['SP1500 CAP'] = spINX['S&P 1500'] # 단위 : 빌리언
new_df['AAPL CAP'] = aaplcap['cap']/1000 # 애플 시가총액 단위 : 밀리언 -> 빌리언 단위로 바꿔줬음
new_df['NewSP1500 CAP'] = new_df['SP1500 CAP'] - new_df['AAPL CAP']
new_df.head()

SP500 CAP  SP400 CAP  SP600 CAP  SP1500 CAP    AAPL CAP  \
date                                                                  
2010-01-04   10088.69     917.16     410.16    11416.01  155.346071   
2010-01-05   10120.19     919.60     408.75    11448.55  155.650970   
2010-01-06   10125.66     924.45     408.36    11458.48  153.160961   
2010-01-07   10166.24     929.12     411.06    11506.43  152.856062   
2010-01-08   10195.52     934.47     412.37    11542.36  153.872393   

            NewSP1500 CAP  
date                       
2010-01-04   11260.663929  
2010-01-05   11292.899030  
2010-01-06   11305.319039  
2010-01-07   11353.573938  
2010-01-08   11388.487607

In [245]:
new_df.tail()

SP500 CAP  SP400 CAP  SP600 CAP  SP1500 CAP    AAPL CAP  \
date                                                                  
2017-12-22   22903.76    1786.09     786.47    25476.32  889.339743   
2017-12-26   22880.12    1787.18     787.42    25454.72  866.777212   
2017-12-27   22898.18    1786.98     787.28    25472.43  866.929662   
2017-12-28   22940.12    1792.38     790.37    25522.87  869.368854   
2017-12-29   22821.24    1782.33     784.39    25387.97  859.967800   

            NewSP1500 CAP  
date                       
2017-12-22   24586.980257  
2017-12-26   24587.942788  
2017-12-27   24605.500338  
2017-12-28   24653.501146  
2017-12-29   24528.002200

In [249]:
new_df['AAPL CAP']['2012-09']

date
2012-09-04    489.972790
2012-09-05    486.568084
2012-09-06    490.938304
2012-09-07    493.987294
2012-09-10    481.130717
2012-09-11    479.555406
2012-09-12    486.212368
2012-09-13    495.816689
2012-09-14    501.813037
2012-09-17    508.012651
2012-09-18    509.537147
2012-09-19    509.689596
2012-09-20    507.199587
2012-09-21    508.215917
2012-09-24    501.457321
2012-09-25    488.956460
2012-09-26    482.909295
2012-09-27    494.597093
2012-09-28    484.281341
Name: AAPL CAP, dtype: float64

In [242]:
len(new_df)

2012

In [244]:
new_df.to_csv(path+'/미국인덱스시가총액(빌리언).csv')